### PR TITLE
Trigger recomputation of personal times when assessment opening and closing timings have been adjusted

### PR DESCRIPTION
**Pull request by zhuhanming**
## Summary

Currently, no recomputation of personal times is being done when an assessment's timings have been shifted, i.e. even if the instructor moved an assessment back to 10 days later, a student might still have their personal opening time set as the previous opening time, which would require the instructor to manually recompute every single's student's personal times.

This fix will now generate a job when a lesson plan item's start_at and end_at are modified, which will then asynchronously update all user's personal times for this specific lesson plan item. No other personal times will be modified.

Some considerations on why only recompute for adjustment of timings and not for other stuff like deletion, etc. :

- We will not run a recomputation when a lesson plan item is deleted, even if it shortens the course. Generally, this shouldn't affect our current estimation of each student's lesson rate and our adjustment of their personal times (to some extent, it doesn't really make sense that an item being deleted = my learning rate suddenly changes). We can lazily update the learning rate upon the student's next submission.
- We will also not run a recomputation when a lesson plan item is created, even if it lengthens the course. Same as before, we can lazily update.
   - However, there may be a case for us to do a recomputation so as to generate a suitable personal time for the new lesson plan item in an eager manner. See #3448.

Other considerations like unsubmission:

- We won't recompute on unsubmission. Although it makes sense that unsubmitting = our previous estimation may no longer be accurate, we must note that unsubmission is done by the instructor, and not the student. As such, it may not make much sense to recompute upon unsubmission.
   - Of course, a counterargument to this is that we're still recomputing upon resubmission, which basically lowers the student's learning rate since now the instructor's response time has been factored into their submission time. This is unfortunately a limitation of our system currently.

## TODO

- Look into whether we should filter for course_users who are students
- Test out with lesson_plan_items that don't affect learning rate currently, e.g. videos
- UX-wise: How can we let the instructor know that there are some students with fixed personal times that won't be affected by their change in timings?

### CODE REVIEW

**Review by ekowidianto — COMMENTED** [review]
Great Thanks! some minor comments below

File: app/jobs/course/lesson_plan/coursewide_personalized_timeline_update_job.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+class Course::LessonPlan::CoursewidePersonalizedTimelineUpdateJob < ApplicationJob
```
[the remark this answers is not in the document]
Comment (zhuhanming) [reply app/jobs/course/lesson_plan/coursewide_personalized_timeline_update_job.rb:2]: I think I will do up a separate PR to build a framework to support the debouncing of jobs.

It doesn't make sense to me to add a new column to the DB everytime we want to do this.

Not to mention that in this scenario, I don't think it's going to happen frequently enough for us to think of it as a major problem.

File: app/controllers/concerns/course/lesson_plan/personalization_concern.rb [hunk 2]
```diff
@@ -28,6 +34,16 @@ def update_personalized_timeline_for(course_user, timeline_algorithm = nil)
...
+    strategy.execute(course_user, precomputed_data, items_to_shift)
+  end
+
+  # Updates the personalized timeline for all course users in the course of the given lesson plan item.
+  # Only the timing for the lesson plan item will be shifted. Generally, you should only call this if the timing of the
+  # lesson plan item has shifted, or other personalized timeline related changes have been made for a specific item.
+  def update_personalized_timeline_for_item(lesson_plan_item)
+    course = lesson_plan_item.course
+    course.course_users.each do |course_user|
+      update_personalized_timeline_for_user(course_user, nil, Set[lesson_plan_item.id])
```
[the remark this answers is not in the document]
Comment (zhuhanming) [reply app/controllers/concerns/course/lesson_plan/personalization_concern.rb:46]: Discussed offline - no need to preload.

File: app/controllers/concerns/course/lesson_plan/personalization_concern.rb [hunk 3]
```diff
@@ -28,6 +34,16 @@ def update_personalized_timeline_for(course_user, timeline_algorithm = nil)
...
-    strategy.execute(course_user, precomputed_data)
+    strategy.execute(course_user, precomputed_data, items_to_shift)
+  end
+
+  # Updates the personalized timeline for all course users in the course of the given lesson plan item.
+  # Only the timing for the lesson plan item will be shifted. Generally, you should only call this if the timing of the
+  # lesson plan item has shifted, or other personalized timeline related changes have been made for a specific item.
+  def update_personalized_timeline_for_item(lesson_plan_item)
+    course = lesson_plan_item.course
+    course.course_users.each do |course_user|
```
[the remark this answers is not in the document]
Comment (zhuhanming) [reply app/controllers/concerns/course/lesson_plan/personalization_concern.rb:45]: Preloaded, thanks for pointing this out!

**Review by ekowidianto — APPROVED** [review]

**Pull request merged**